File: LeetCode/Hard/2163-Minimum_Difference_In_Sums_After_Removal_Of_Elements/2163-MinimumDifferenceInSumsAfterRemovalOfElements.cpp
#include <vector>
#include <queue>
#include <numeric>
#include <algorithm>

class Solution
{
public:
    long long minimumDifference(vector<int> &nums)
    {
        int n3 = nums.size();
        int n = n3 / 3;

        std::priority_queue<int, std::vector<int>, std::greater<int>> min_heap;
        std::vector<long long> suffix_sum(n3, 0);
        long long current_sum_suffix = 0;

        for (int i = n3 - 1; i >= n - 1; --i)
        {
            min_heap.push(nums[i]);
            current_sum_suffix += nums[i];

            if (min_heap.size() > n)
            {
                current_sum_suffix -= min_heap.top();
                min_heap.pop();
            }

            if (min_heap.size() == n)
            {
                suffix_sum[i] = current_sum_suffix;
            }
        }

        std::priority_queue<int> max_heap_prefix;
        long long current_sum_prefix = 0;
        long long min_diff = std::numeric_limits<long long>::max();

        for (int i = 0; i < 2 * n; ++i)
        {
            max_heap_prefix.push(nums[i]);
            current_sum_prefix += nums[i];

            if (max_heap_prefix.size() > n)
            {
                current_sum_prefix -= max_heap_prefix.top();
                max_heap_prefix.pop();
            }

            if (max_heap_prefix.size() == n)
            {
                min_diff = std::min(min_diff, current_sum_prefix - suffix_sum[i + 1]);
            }
        }

        return min_diff;
    }
};

File: LeetCode/Hard/2163-Minimum_Difference_In_Sums_After_Removal_Of_Elements/2163-MinimumDifferenceInSumsAfterRemovalOfElements.js
class HeapQueue {
    constructor(cmp) {
        this.data = [];
        this.cmp = cmp;
    }

    push(value) {
        this.data.push(value);
        let i = this.data.length - 1;

        while (i > 0) {
            const parent = (i - 1) >> 1;
            if (this.cmp(this.data[i], this.data[parent]) < 0) {
                [this.data[i], this.data[parent]] = [this.data[parent], this.data[i]];
                i = parent;
            } else
                break;
        }
    }

    pop() {
        if (this.isEmpty()) {
            return undefined;
        }
        if (this.data.length === 1) {
            return this.data.pop();
        }

        const top = this.data[0];
        this.data[0] = this.data.pop();
        this._heapifyDown(0);
        return top;
    }

    _heapifyDown(i) {
        const n = this.data.length;

        while (true) {
            let l = 2 * i + 1;
            let r = 2 * i + 2;
            let minI = i;

            if (l < n && this.cmp(this.data[l], this.data[minI]) < 0)
                minI = l;
            if (r < n && this.cmp(this.data[r], this.data[minI]) < 0)
                minI = r;
            if (minI !== i) {
                [this.data[i], this.data[minI]] = [this.data[minI], this.data[i]]
                i = minI;
            } else
                break;
        }
    }

    peek() {
        if (this.isEmpty())
            return undefined;
        return this.data[0];
    }

    isEmpty() {
        return this.data.length === 0;
    }

    size() {
        return this.data.length
    }
}

/**
 * @param {number[]} nums
 * @return {number}
 */
var minimumDifference = function (nums) {
    const n3 = nums.length;
    const n = Math.floor(n3 / 3);

    const min_heap = new HeapQueue((a, b) => a - b);
    const suffix_sum = new Array(n3).fill(0);
    let curr_sum = 0;

    for (let i = n3 - 1; i >= n - 1; --i) {
        min_heap.push(nums[i]);
        curr_sum += nums[i];

        if (min_heap.size() > n) {
            curr_sum -= min_heap.pop();
        }

        if (min_heap.size() === n) {
            suffix_sum[i] = curr_sum;
        }
    }

    const max_heap = new HeapQueue((a, b) => b - a);
    curr_sum = 0
    let ans = Infinity;

    for (let i = 0; i < 2 * n; ++i) {
        max_heap.push(nums[i]);
        curr_sum += nums[i];

        if (max_heap.size() > n) {
            curr_sum -= max_heap.pop();
        }

        if (max_heap.size() === n) {
            ans = Math.min(ans, curr_sum - suffix_sum[i + 1]);
        }
    }

    return ans;
};